diff --git a/server/scsynth/SC_World.cpp b/server/scsynth/SC_World.cpp
index 25682d4..aa34079 100644
--- a/server/scsynth/SC_World.cpp
+++ b/server/scsynth/SC_World.cpp
@@ -255,6 +255,15 @@ void InterfaceTable_Init() {
 void initialize_library(const char* mUGensPluginPath);
 void initializeScheduler();
 
+static bool gLibInitted = false;
+
+void World_UnloadPlugins() {
+    if (!gLibInitted)
+        return;
+    deinitialize_library();
+    gLibInitted = false;
+}
+
 static void World_LoadGraphDefs(World* world);
 void World_LoadGraphDefs(World* world) {
     GraphDef* list = nullptr;
@@ -313,7 +322,6 @@ World* World_New(WorldOptions* inOptions) {
     World* world = nullptr;
 
     try {
-        static bool gLibInitted = false;
         if (!gLibInitted) {
             InterfaceTable_Init();
             initialize_library(inOptions->mUGensPluginPath);
@@ -931,7 +939,7 @@ void World_Cleanup(World* world, bool unload_plugins) {
     }
 
     if (unload_plugins)
-        deinitialize_library();
+        World_UnloadPlugins();
 
     HiddenWorld* hw = world->hw;
 
